File: Algorithms/Pattern/Recursion/sum.js
/**
 * 使用递归实现数组求和
 * @param {Number[]} arr
 */
function sum(arr) {
  return _sum(arr, 0)

  function _sum(arr, index) {
    // 递归终止条件
    if (index === arr.length) {
      return 0
    }

    return arr[index] + _sum(arr, index + 1) // 递归
  }
}

// test

arr = [1, 2, 3, 4, 5, 6, 7, 8]

const sums = sum(arr)

console.log(sums)

File: Java/leetcode/203/Sum.java
/**
 * Sum 使用递归实现数组求和
 */
public class Sum {

  public static int sum(int[] arr) {
    return sum(arr, 0);
  }

  private static int sum(int[] arr, int index) {
    if (index == arr.length) {
      return 0;
    }

    int res = arr[index] + sum(arr, index + 1); // 单步调试变量
    return res;
  }

  public static void main(String[] args) {

    int[] nums = { 1, 2, 3, 4, 5, 6, 7, 8 };
    System.out.println(sum(nums));

  }

}